runtime/io/common/src/aws/smithy/kotlin/runtime/io/HashingByteReadChannel.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.hashing.HashFunction
+import aws.smithy.kotlin.runtime.io.internal.commonPeek
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * A channel which hashes data as it is being read
+ * @param hash The [HashFunction] to hash data with
+ * @param chan the [SdkByteReadChannel] to hash
+ */
+@InternalApi
+public class HashingByteReadChannel(private val hash: HashFunction, private val chan: SdkByteReadChannel) : SdkByteReadChannel by chan {
+    public override suspend fun read(sink: SdkBuffer, limit: Long): Long {
+        val buffer = SdkBuffer()
+
+        val rc = chan.read(buffer, limit)
+        if (rc == -1L) { return rc } // got 0 bytes, no need to hash anything
+
+        val dataToHash = buffer.commonPeek().readByteArray()
```

Comment:
`peek` should work. The code-completion / suggestions in my IDE don't work great with the new I/O refactor, and I only picked `commonPeek` because it showed up and `peek` didn't.

Reply:
Not sure it's the I/O refactor, my IDE is having trouble with completions as well FWIW.

Reply:
Hmm, I only ever noticed it with the newer I/O stuff. This is an example of what I mean:

<img width="816" alt="Screen Shot 2023-01-20 at 2 44 40 PM" src="https://user-images.githubusercontent.com/30608308/213801932-8ee4315c-4edd-4893-a12b-1b1a202d89a5.png">